### PR TITLE
Update `raw.hpp` so that we can serialize and deserialize `const` types

### LINKED ISSUE #796: Update `raw.hpp` so that we can serialize and deserialize `const` types
Some types, such as `block_state_legacy::action_mroot_savanna`, are only set in the constructor and never modified. So they are logically const and could be accessed from multiple threads safely. 

Using this would allow us to remove some friend declarations (we use `private` as a way to prevent unchecked access to non thread-safe members).

Ideally we would make the member type const. 
So for `block_state_legacy::action_mroot_savanna` the type would be:
`const std::optional<digest_type>` instead of 
`std::optional<digest_type>`.

Again, the serialization is what trips us up. I feel that the unpack functions in `raw.hpp` should accept to unpack into const types (logically deserializing to recreate a value which is never modified after creation). Right now we cannot mark anything that needs to be serialized const.

So for `optional<T>` the signature should be: `void unpack( Stream& s, const std::optional<T>& v )`

even if it requires a `const_cast` internally.

### PR DESCRIPTION
**Pull request by greg7mdp**
Resolves #796.

Some types are only set in the constructor and never modified. So they are logically const and could be accessed from multiple threads safely.

Using this would allow us to remove some friend declarations (we use private as a way to prevent unchecked access to non thread-safe members).

Ideally we would make the member type const.
The serialization is what trips us up. I feel that the unpack functions in raw.hpp should accept to unpack const members of structs (logically deserializing to recreate a value which is never modified after creation). Right now we cannot mark non-funadamental types that needs to be serialized const.

Also added a test.

### CODE REVIEW

**Review by heifner — CHANGES_REQUESTED** [review]
Since we have the time. What about making a pass through the code and making what can be const, `const`.

**Comment by greg7mdp** [comment]
> Since we have the time. What about making a pass through the code and making what can be const, `const`.

I started doing that, but it does require a lot of changes through the code. For example most of the `block_header_state` should be const. But typically we create the empty struct and assign members after the fact (see [here](https://github.com/AntelopeIO/spring/blob/8cc763f3c331aff2287940669c3ca1d60dab2285/libraries/chain/block_state.cpp#L136-L160) for example). Also the `finish_next` function.  

Probably if we want to do this I should create another issue.

**Comment by ericpassmore** [comment]
Note:start
category: Chores
component: Internal
summary: Update `raw.hpp` so that we can serialize and deserialize `const` types.
Note:end

File: libraries/chain/include/eosio/chain/finality/finalizer.hpp [hunk 1]
```diff
@@ -184,7 +189,13 @@ namespace eosio::chain {
 
       // for testing purposes only, not thread safe
       const fsi_t& get_fsi(const bls_public_key& k) const { return finalizers.at(k).fsi; }
-      void         set_fsi(const bls_public_key& k, const fsi_t& fsi) { finalizers[k].fsi = fsi; }
+
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/finality/finalizer.hpp:192]: I think remove this blank line or add another comment saying this is only for testing purposes and not thread safe.

**Review by heifner — APPROVED** [review]

File: libraries/chain/include/eosio/chain/finality/finalizer.hpp (hunk 1 above)
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/finality/finalizer.hpp:192]: removed!

**Review by heifner — APPROVED** [review]

**Review by linh2931 — APPROVED** [review]

**Pull request merged**